src/vos/vos_space.c

```diff
@@ -0,0 +1,344 @@
+		if (iod->iod_type == DAOS_IOD_SINGLE) {
+			size = iod->iod_size;
+			media = vos_media_select(pool, iod->iod_type, size);
+
+			/* Single value record */
+			if (media == DAOS_MEDIA_SCM) {
+				scm += vos_recx2irec_size(size, csums);
+			} else {
+				scm += vos_recx2irec_size(0, csums);
+				if (iod->iod_size != 0)
+					nvme += vos_byte2blkcnt(iod->iod_size);
+			}
+			/* Assume one more SV tree node created */
+			scm += umem_slab_usize(umm, VOS_SLAB_SV_NODE);
+			continue;
+		}
+
+		/* Array value */
+		for (j = 0; j < iod->iod_nr; j++) {
+			recx = &iod->iod_recxs[j];
+			recx_csum = csums ? &csums[j] : NULL;
+
+			size = recx->rx_nr * iod->iod_size;
+			media = vos_media_select(pool, iod->iod_type, size);
+
+			/* Extent */
+			if (media == DAOS_MEDIA_SCM)
+				scm += size;
+			else if (size != 0)
+				nvme += vos_byte2blkcnt(size);
+			/* EVT desc */
+			scm += umem_slab_usize(umm, VOS_SLAB_EVT_DESC);
+			/* Checksum */
+			scm += recx_csum_len(recx, recx_csum, iod->iod_size);
+			/* Assume one more evtree node created */
+			scm += umem_slab_usize(umm, VOS_SLAB_EVT_NODE);
+		}
+	}
+
+	space_est[DAOS_MEDIA_SCM] = scm;
+	space_est[DAOS_MEDIA_NVME] = nvme * VOS_BLK_SZ;
+}
+
+int
+vos_space_hold(struct vos_pool *pool, daos_key_t *dkey, unsigned int iod_nr,
+	       daos_iod_t *iods, struct dcs_iod_csums *iods_csums,
+	       daos_size_t *space_hld)
+{
+	struct vos_pool_space	vps = { 0 };
+	daos_size_t		space_est[DAOS_MEDIA_MAX] = { 0, 0 };
+	daos_size_t		scm_left, nvme_left;
+	int			rc;
+
+	rc = vos_space_query(pool, &vps, false);
+	if (rc) {
+		D_ERROR("Query pool:"DF_UUID" space failed. "DF_RC"\n",
+			DP_UUID(pool->vp_id), DP_RC(rc));
+		return rc;
+	}
+
+	scm_left = SCM_FREE(&vps);
+	if (scm_left < SCM_SYS(&vps))
+		goto error;
+
+	scm_left -= SCM_SYS(&vps);
+	if (scm_left < POOL_SCM_HELD(pool))
+		goto error;
+
+	scm_left -= POOL_SCM_HELD(pool);
+
+	nvme_left = NVME_FREE(&vps);
+	if (pool->vp_vea_info) {
+		if (nvme_left < NVME_SYS(&vps))
+			goto error;
+		nvme_left -= NVME_SYS(&vps);
+		/* 'NVMe held' has already been excluded from 'NVMe free' */
+	}
+
+	estimate_space(pool, dkey, iod_nr, iods, iods_csums, &space_est[0]);
+
+	if (scm_left < space_est[DAOS_MEDIA_SCM])
+		goto error;
+
+	if (pool->vp_vea_info) {
+		if (nvme_left < space_est[DAOS_MEDIA_NVME])
+			goto error;
+	}
+
+	space_hld[DAOS_MEDIA_SCM]	= space_est[DAOS_MEDIA_SCM];
+	space_hld[DAOS_MEDIA_NVME]	= space_est[DAOS_MEDIA_NVME];
+	POOL_SCM_HELD(pool)		+= space_hld[DAOS_MEDIA_SCM];
+	POOL_NVME_HELD(pool)		+= space_hld[DAOS_MEDIA_NVME];
+
+	return 0;
+error:
+	D_ERROR("Pool:"DF_UUID" is full. SCM: free["DF_U64"], sys["DF_U64"], "
+		"hld["DF_U64"], est["DF_U64"] NVMe: free["DF_U64"], "
+		"sys["DF_U64"], hld["DF_U64"], est["DF_U64"]\n",
+		DP_UUID(pool->vp_id), SCM_FREE(&vps), SCM_SYS(&vps),
+		POOL_SCM_HELD(pool), space_est[DAOS_MEDIA_SCM],
+		NVME_FREE(&vps), NVME_SYS(&vps), POOL_NVME_HELD(pool),
+		space_est[DAOS_MEDIA_NVME]);
+
+	return -DER_NOSPACE;
```

Comment:
If we return -DER_NOSPACE, then the client/application will be aware the failure and quite possible fall out. But under such case, since we still have some system reserved space, should we trigger VOS aggregation and ask client to retry the update sometime later (via -DER_AGAIN or similar) without let the application to know the space trouble?

On the other hand, the space reservation is based on the worst estimation, the real real space usage may not need so much. If a lot of current updates hold (but not real used) too much space, then it may cause others. So can we make the best estimation with min consumed space. If it is not enough during the real update, then report -DER_NOSPACE at that time?

Reply:
This patch only implement space reserving in VOS. To handle ENOSPACE gracefully, we need to improve iosrv scheduler to prioritize GC & Aggregation when pool is under space pressure, and return -ENOSPACE to client only when space reclaim is done by GC & aggregation, and probably return -EINPROGRESS for update while waiting GC and aggregation done? All these will be implemented in follow-on patches.

For the second case you mentioned, I think returning -DER_NOSPACE to client is the best thing we can do, it's not accurate, but we'd try to ensure the sys space is not encroached by client update.